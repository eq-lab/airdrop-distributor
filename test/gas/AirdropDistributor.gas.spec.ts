import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import snapshotGasCost from '@uniswap/snapshot-gas-cost';
import { ethers } from 'hardhat';
import { createAirdrop } from '../shared/fixtures';
import { constructMerkleTree, generateRandomAirdropData } from '../shared/utils';
import { parseUnits } from 'ethers';

describe('AirdropDistributor gas', () => {
  it('claim gas 100 eligible users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);
    const [_, user] = await ethers.getSigners();

    const airdropData = generateRandomAirdropData(99);
    const userAmount = parseUnits('100', 18);
    airdropData.push({ address: user.address, amount: userAmount });

    const tree = constructMerkleTree(airdropData);
    await airdropDistributor.connect(owner).updateRoot(tree.root);

    const userProof = tree.getProof([user.address, userAmount]);
    const txReceipt = await (await airdropDistributor.connect(user).claim(userAmount, userProof)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('claim gas 1000 eligible users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);
    const [_, user] = await ethers.getSigners();

    const airdropData = generateRandomAirdropData(999);
    const userAmount = parseUnits('100', 18);
    airdropData.push({ address: user.address, amount: userAmount });

    const tree = constructMerkleTree(airdropData);
    await airdropDistributor.connect(owner).updateRoot(tree.root);

    const userProof = tree.getProof([user.address, userAmount]);
    const txReceipt = await (await airdropDistributor.connect(user).claim(userAmount, userProof)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('claim gas 10000 eligible users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);
    const [_, user] = await ethers.getSigners();

    const airdropData = generateRandomAirdropData(9999);
    const userAmount = parseUnits('100', 18);
    airdropData.push({ address: user.address, amount: userAmount });

    const tree = constructMerkleTree(airdropData);
    await airdropDistributor.connect(owner).updateRoot(tree.root);

    const userProof = tree.getProof([user.address, userAmount]);
    const txReceipt = await (await airdropDistributor.connect(user).claim(userAmount, userProof)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('claim gas 50000 eligible users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);
    const [_, user] = await ethers.getSigners();

    const airdropData = generateRandomAirdropData(49999);
    const userAmount = parseUnits('100', 18);
    airdropData.push({ address: user.address, amount: userAmount });

    const tree = constructMerkleTree(airdropData);
    await airdropDistributor.connect(owner).updateRoot(tree.root);

    const userProof = tree.getProof([user.address, userAmount]);
    const txReceipt = await (await airdropDistributor.connect(user).claim(userAmount, userProof)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('updateRoot gas 100 users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(100);
    const tree = constructMerkleTree(airdropData);
    const txReceipt = await (await airdropDistributor.connect(owner).updateRoot(tree.root)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('updateRoot gas 1000 users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(1000);
    const tree = constructMerkleTree(airdropData);
    const txReceipt = await (await airdropDistributor.connect(owner).updateRoot(tree.root)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('updateRoot gas 10000 users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(10000);
    const tree = constructMerkleTree(airdropData);
    const txReceipt = await (await airdropDistributor.connect(owner).updateRoot(tree.root)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });

  it('updateRoot gas 50000 users', async () => {
    const { airdropDistributor, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(50000);
    const tree = constructMerkleTree(airdropData);
    const txReceipt = await (await airdropDistributor.connect(owner).updateRoot(tree.root)).wait();
    await snapshotGasCost(Number(txReceipt?.gasUsed));
  });
});
