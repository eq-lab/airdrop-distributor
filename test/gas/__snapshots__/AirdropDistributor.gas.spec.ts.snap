// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirdropDistributor gas claim gas 100 eligible users 1`] = `
Object {
  "calldataByteLength": 324,
  "gasUsed": 95574,
}
`;

exports[`AirdropDistributor gas claim gas 1000 eligible users 1`] = `
Object {
  "calldataByteLength": 420,
  "gasUsed": 98445,
}
`;

exports[`AirdropDistributor gas claim gas 10000 eligible users 1`] = `
Object {
  "calldataByteLength": 548,
  "gasUsed": 102079,
}
`;

exports[`AirdropDistributor gas claim gas 50000 eligible users 1`] = `
Object {
  "calldataByteLength": 612,
  "gasUsed": 103953,
}
`;

exports[`AirdropDistributor gas updateRoot gas 100 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47129,
}
`;

exports[`AirdropDistributor gas updateRoot gas 1000 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;

exports[`AirdropDistributor gas updateRoot gas 10000 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;

exports[`AirdropDistributor gas updateRoot gas 50000 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;
